src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs

```diff
@@ -0,0 +1,415 @@
+                    return null;
+                }
+                set
+                {
+                    if (Fields.ContainsKey(STRING_DATA_PROPERTY_NAME))
+                    {
+                        Fields[STRING_DATA_PROPERTY_NAME].S = value;
+                    }
+                    else
+                    {
+                        Fields.Add(STRING_DATA_PROPERTY_NAME, new AttributeValue(value));
+                    }
+                }
+            }
+        }
+
+        private class GrainStateTableDataManager
+        {
+            public string TableName { get; private set; }
+            private readonly Logger logger;
+            private readonly DynamoDBDataManager<GrainStateRecord> tableManager;
+
+            public GrainStateTableDataManager(string tableName, string accessKey, string secretKey, string service, int readCapacityUnits, int writeCapacityUnits, Logger logger)
+            {
+                this.logger = logger;
+                TableName = tableName;
+                tableManager = new DynamoDBDataManager<GrainStateRecord>(tableName, accessKey, secretKey, service, readCapacityUnits, writeCapacityUnits, logger);
+            }
+
+            public Task InitTableAsync()
+            {
+                return tableManager.InitTableAsync();
+            }
+
+            public Task<GrainStateRecord> Read(string partitionKey, string rowKey)
+            {
+                if (logger.IsVerbose3) logger.Verbose3(ErrorCode.StorageProviderBase, "Reading: PartitionKey={0} RowKey={1} from Table={2}", partitionKey, rowKey, TableName);
+
+                try
+                {
+                    var data = tableManager.ReadSingleTableEntryAsync(partitionKey, rowKey);
+                    if (data == null && logger.IsVerbose2)
+                    {
+                        logger.Verbose2(ErrorCode.StorageProviderBase, "DataNotFound reading: PartitionKey={0} RowKey={1} from Table={2}", partitionKey, rowKey, TableName);
+                    }
+                    return data;
+                }
+                catch (Exception exc)
+                {
+                    if (logger.IsVerbose2) logger.Verbose2(ErrorCode.StorageProviderBase, "Failure reading data (exception): PartitionKey={0} RowKey={1} from Table={2} Exception={3}", partitionKey, rowKey, TableName, LogFormatter.PrintException(exc));
+                    throw;
+                }
+            }
+
+            public async Task Write(GrainStateRecord record)
+            {
+                if (logger.IsVerbose3) logger.Verbose3(ErrorCode.StorageProviderBase, "Writing: PartitionKey={0} RowKey={1} to Table={2} with ETag={3}", record.PartitionKey, record.RowKey, TableName, record.ETag);
+                string eTag = record.ETag == 0 ?
+                    await tableManager.CreateTableEntryAsync(record).ConfigureAwait(false) :
+                    await tableManager.UpdateTableEntryAsync(record, record.ETag.ToString()).ConfigureAwait(false);
```

Comment:
You have to wrap this with a try and catch and rethrow an `InconsistentStateException` when the write failed because the row already exists or there was a version conflict. Looking at the Azure code, this isn't implemented there, which is a bug.

a note for @veikkoeeva : having the insert and update separate will simplify the db scripts greatly (no need to lock anything). I don't know how I didn't think of it before.